util/webhook/webhook.go

```diff
@@ -390,11 +391,33 @@ func sourceRevisionHasChanged(source v1alpha1.ApplicationSource, revision string
 	targetRevisionHasPrefixList := []string{"refs/heads/", "refs/tags/"}
 	for _, prefix := range targetRevisionHasPrefixList {
 		if strings.HasPrefix(source.TargetRevision, prefix) {
-			return revision == targetRev
+			return compareRevisions(revision, targetRev)
 		}
 	}
 
-	return source.TargetRevision == revision
+	return compareRevisions(revision, source.TargetRevision)
+}
+
+func compareRevisions(revision string, targetRevision string) bool {
+	if revision == targetRevision {
+		return true
+	}
+
+	// If basic equality checking fails, it might be that the target revision is
+	// a semver version constraint
+	constraint, err := semver.NewConstraint(targetRevision)
+	if err != nil {
+		// The target revision is not a constraint
+		return false
+	}
+
+	version, err := semver.NewVersion(revision)
+	if err != nil {
+		// The new revision is not a valid semver version, so it can't match the constraint.
+		return false
+	}
+
+	return constraint.Check(version)
```

Comment:
I suppose in theory this logic could trigger an unnecessary sync if the new tag is not the latest tag that would match the constraint.

For example, if the constraint was "1.1.*", there was already a tag "1.1.5" which the application was already synced to, and somebody introduces (or moves) a tag "1.1.4". A sync would be triggered, but the outcome would still be that the application would select the "1.1.5" revision.

However, I doubt this is a circumstance that would happen for any sensible SemVer tagging strategy.